tedana/decomposition/gift_pca.py

```diff
@@ -0,0 +1,599 @@
+        v2 = np.hstack((np.arange(1, n[1] + 1), np.arange(n[1] - 1, 0,
+                                                          -1)))[np.newaxis, :]
+
+        vd = np.dot(v1.T, v2)
+        xc = xc / vd
+        parzen_window_2D = np.dot(parzen_w_1, parzen_w_2.T)
+        xc = xc * parzen_window_2D
+        xf = fftshift(fft2(xc))
+
+    elif x.ndim == 3 and min(n) != 1:  # 3D
+        xc = np.zeros((2 * n[0] - 1, 2 * n[1] - 1, 2 * n[2] - 1))
+        for m3 in range(n[2] - 1):
+            temp = np.zeros((2 * n[0] - 1, 2 * n[1] - 1))
+            for k in range(n[2] - m3):
+                temp = temp + correlate2d(x[:, :, k + m3], x[:, :, k])
+                # default option:
+                # computes raw correlations with NO normalization
+                # -- Matlab help on xcorr
+            xc[:, :, (n[2] - 1) - m3] = temp
+            xc[:, :, (n[2] - 1) + m3] = temp
+
+        # Bias correction
+        v1 = np.hstack((np.arange(1, n[0] + 1), np.arange(n[0] - 1, 0,
+                                                          -1)))[np.newaxis, :]
+        v2 = np.hstack((np.arange(1, n[1] + 1), np.arange(n[1] - 1, 0,
+                                                          -1)))[np.newaxis, :]
+        v3 = np.arange(n[2], 0, -1)
+
+        vd = np.dot(v1.T, v2)
+        vcu = np.zeros((2 * n[0] - 1, 2 * n[1] - 1, 2 * n[2] - 1))
+        for m3 in range(n[2]):
+            vcu[:, :, (n[2] - 1) - m3] = vd * v3[m3]
+            vcu[:, :, (n[2] - 1) + m3] = vd * v3[m3]
+
+        # Possible source of NAN values
+        xc = xc / vcu
+
+        parzen_window_2D = np.dot(parzen_w_1[np.newaxis, :].T,
+                                  parzen_w_2[np.newaxis, :])
+        parzen_window_3D = np.zeros((2 * n[0] - 1, 2 * n[1] - 1, 2 * n[2] - 1))
+        for m3 in range(n[2] - 1):
+            parzen_window_3D[:, :, (n[2] - 1) - m3] = np.dot(
+                parzen_window_2D, parzen_w_3[n[2] - 1 - m3])
+            parzen_window_3D[:, :, (n[2] - 1) + m3] = np.dot(
+                parzen_window_2D, parzen_w_3[n[2] - 1 + m3])
+
+        xc = xc * parzen_window_3D
+
+        xf = fftshift(fftn(xc))
+
+    else:
+
+        raise ValueError('Unrecognized matrix dimension.')
+
+    xf = abs(xf)
+    xf[xf < 1e-4] = 1e-4
+    out = 0.5 * np.log(2 * np.pi * np.exp(1)) + _sumN(np.log(abs(
+        (xf)))) / 2 / _sumN(abs(xf))
+
+    return out
+
+
+def _est_indp_sp(x):
+    """
+    Estimate the effective number of independent samples based on the maximum
+    entropy rate principle of stationary random process.
+
+    Parameters
+    ----------
+    x :
+
+    Returns
+    -------
+    s :
+    entrate_m :
+    """
+
+    dimv = x.shape
+    s0 = 0
+
+    for j in range(np.min(dimv) - 1):
+        x_sb = _subsampling(x, j + 1, [0, 0, 0])
+        if j == 0:
+            LGR.info(
+                'Estimating the entropy rate of the Gaussian component with subsampling depth {},'
+                .format(j))
+        else:
```

Comment:
This is just a pseudo progress bar ? I think I'd rather add an actual progress bar, at that point. But how long does this step really take ?

Reply:
For what I've seen, it takes 2 iterations. It is a progress bar but still it's meant to show how many iterations were needed. I say we just print it when the loop is done. What do you think?

Reply:
I like that !